Fix WebUI removing parameters from magnet links

Piccirello opened the pull request:
List of supported parameters retrieved from [here](https://en.wikipedia.org/wiki/Magnet_URI_scheme)

Closes #10939

glassez commented:
@Piccirello, please describe the problem you solve in this PR.

Piccirello commented:
@glassez have you checked out the issue I linked to? The problem is described there, but let me know if that’s not clear enough.

glassez commented:
>The problem is described there, but let me know if that’s not clear enough.

I just want to know how is Magnet URI handled by WebUI. Desktop app accepts entire Magnet URI as a string. IIRC, WebUI does the job in a different way.

Piccirello commented on src/webui/www/private/scripts/client.js:
```diff
@@ -828,12 +828,28 @@ function registerMagnetHandler() {
-    // Remove the download param from WebUI URL hash
     delete hashParams.download;
+
+    const magnetLinkParams = new Set(['dn', 'xl', 'xt', 'as', 'xs', 'kt', 'mt', 'tr']);
```
Sorted alphabetically

Piccirello commented:
Addressed PR comments

@glassez The WebUI send the entire magnet link as entered on the upload page. It doesn't do anything special with it. In this case, we're receiving the link from an untrusted source, so we perform some basic parsing.

glassez commented:
>so we perform some basic parsing.

IIRC, you drop unknown params from Magnet URI. Why not drop invalid Magnet URI at all?

Piccirello commented:
Per the magnet uri spec, there may be additional, application-specific parameters. We shouldn't drop the uri because we don't support the parameter, but we also shouldn't pass it along.

glassez commented:
>Per the magnet uri spec, there may be additional, application-specific parameters. We shouldn't drop the uri because we don't support the parameter, but we also shouldn't pass it along.

Wouldn't it be better to let libtorrent handle it? Maybe there are some libtorrent-specific parameters?

Chocobo1 commented on src/webui/www/private/scripts/mocha-init.js:
```diff
@@ -126,8 +126,11 @@ var initializeWindows = function() {
-        if (urls && urls.length)
-            contentUrl += '?urls=' + urls.join("|");
+        if (urls && (urls.length > 0)) {
+            contentUrl += '?urls=' + urls.map(function(url) {
```
parentheses: `contentUrl += ('?urls=' + urls.map(function(url) {...}));`

Chocobo1 commented:
>Wouldn't it be better to let libtorrent handle it? Maybe there are some libtorrent-specific parameters?

I was thinking about it too... my instinct was to pass the original magnet URL as is without any filtering.

Piccirello commented:
I've changed the code to remove any filtering outside of checking if there's a download url specified.

Chocobo1 commented:
we have conflicts now.

Piccirello commented:
Rebased

Chocobo1 commented:
@Piccirello 
Thanks!
